### DISCUSSION TITLE
CPython compatibility

### DISCUSSION THREAD

**Discussion by marekmosiewicz** [oracle/graal#2970] [Ideas] [answered]
Hi, I would like to know if c extensions to Python are planned to support. I just have crazy idea to compile Odoo to Graal (https://github.com/marekmosiewicz/goblet) . I need PostgreSQL DB API (psycopg2) which uses C. Maybe it is also possible to replace it with some DB API compilant wrapper around JDBC

**Comment by timfel** [comment, marked as answer]
Support for Python C extension has been a goal from the start, but we've been focusing on data science and ML tasks, so NumPy/SciPy/SciKit-Learn and related packages, which do work to some degree. We have not tried many other C extensions, but we welcome other people trying and sending us reports and issues :-) 

I am not sure what running Odoo entails, but note that we have very little support for anything involving networking (only basic and very slow socket support, no SSL) - it's on the roadmap, but not won't be done before the latter half of next year.

**Comment by marekmosiewicz** [comment]
I have succesfuly created scripts to install odoo and graalpython from
Gradle.

Graalpython launched odoo but of course it complains abut missing
libraries (PyPDF is the first)
If I have some time meybe I will look more into this.

Network is not so much problem as Odoo is web application and
as it seems to be is written with Werkzeug which is even available
as graalpython lib already.

What is more difficult is psycopg2 lib to connect with PostgreSQL.
That is written as C extenstion.

I could try to compile it however I do not know how.
Should I do it with LLVM or some how in Graal directly ?

Cheers,
    Marek Mosiewicz
    http://marekmosiewicz.pl

W dniu pią, 13.11.2020 o godzinie 06∶29 -0800, użytkownik Tim
Felgentreff napisał:

**Comment by timfel** [comment]
You just build the extension as normal either using `graalpython setup.py build` or `pip install .` or some variant thereof. It automatically builds with the correct compiler
